### PR TITLE
Enhancement of Websocket connector

### PR DESCRIPTION
**Pull request by milanbhagwat**
1) Add support for string type data,
2) Add support of Path/Query Params and headers Parsing

### CODE REVIEW

File: trigger/wsserver/descriptor.json [hunk 1]
```diff
@@ -41,18 +41,19 @@
...
...cept-Charset\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Accept-Encoding\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Content-Type\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Content-Length\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Connection\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Cookie\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Pragma\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Sec-Websocket-Key\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Sec-Websocket-Version\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false},{\"parameterName\":\"Upgrade\",\"type\":\"string\",\"repeating\":\"false\",\"required\":\"false\",\"visible\":false}]"
```
Comment (lixingwang) [inline trigger/wsserver/descriptor.json]: This is invalid from the OSS perspective, please remove value or change to properly one

**Review by lixingwang — CHANGES_REQUESTED** [review]
I think we also should think about adding WebSocket Ping/Pong.

File: trigger/wsserver/descriptor.json [hunk 2]
```diff
@@ -41,18 +41,19 @@
   "output": [
     {
       "name": "pathParams",
-      "type": "params",
+      "type": "object",
```
Comment (vijaynalawade) [inline trigger/wsserver/descriptor.json]: This should be params.

File: trigger/wsserver/descriptor.json (hunk 1 above)
Comment (vijaynalawade) [inline trigger/wsserver/descriptor.json]: It is incorrect value based on the type.

**Review by pointlander — APPROVED** [review]

File: activity/ws/activity.go [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
+	"encoding/base64"
 	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+
+	//"net/url"
```
Comment (lixingwang) [inline activity/ws/activity.go]: delete it

File: activity/ws/activity.go [hunk 4]
```diff
@@ -48,25 +57,214 @@ func (a *Activity) Metadata() *activity.Metadata {
...
+			tlsconfig := &tls.Config{}
+			allowInsecure := a.settings.AllowInsecure
+			if allowInsecure {
+				tlsconfig.InsecureSkipVerify = true
+			} else {
+				// identify if OSS
+				var cacertObj map[string]interface{}
+				if a.settings.CaCert != "" {
+					err = json.Unmarshal([]byte(a.settings.CaCert), &cacertObj)
+					if err != nil { //OSS way
```
Comment (lixingwang) [inline activity/ws/activity.go]: delete OSS/flogo those kind of comments

File: activity/ws/activity.go [hunk 5]
```diff
@@ -48,25 +57,214 @@ func (a *Activity) Metadata() *activity.Metadata {
...
+			dialer = websocket.Dialer{TLSClientConfig: tlsconfig}
+		} else {
+			dialer = *websocket.DefaultDialer
+		}
+		ctx.Logger().Info("Creating new connection")
+		conn, res, err := dialer.Dial(builtURL, h)
+		if err != nil {
+			if res != nil {
+				defer res.Body.Close()
+				body, err := ioutil.ReadAll(res.Body)
```
Comment (lixingwang) [inline activity/ws/activity.go]: we should check err just in case

File: activity/ws/metadata.go [hunk 6]
```diff
@@ -1,25 +1,47 @@
...
+	CaCert        string `md:"caCert"`
 }
 
 // Input is the input into the websocket proxy
 type Input struct {
-	Message interface{} `md:"message"`
+	Message     interface{}            `md:"message,required"`
+	PathParams  map[string]string      `md:"pathParams"`
+	QueryParams map[string]interface{} `md:"queryParams"`
+	Headers     map[string]interface{} `md:"headers"`
```
Comment (lixingwang) [inline activity/ws/metadata.go]: As we have params type for Query and Headers, so let's correct Go type tp map[stsring]string

File: go.mod [hunk 7]
```diff
@@ -2,10 +2,13 @@ module github.com/project-flogo/websocket
...
 require (
 	github.com/gorilla/websocket v1.4.0
-	github.com/julienschmidt/httprouter v1.2.0
-	github.com/project-flogo/contrib/activity/rest v0.9.1-0.20190702155437-52b965a4fff5
-	github.com/project-flogo/contrib/trigger/rest v0.9.1-0.20190702155437-52b965a4fff5
-	github.com/project-flogo/core v0.9.3-0.20190626212829-3b488c2791d6
-	github.com/project-flogo/microgateway v0.0.0-20190708190753-c54f135979ec
-	github.com/stretchr/testify v1.3.0
+	github.com/julienschmidt/httprouter v1.3.0
+	github.com/project-flogo/contrib/activity/rest v0.10.0
```
Comment (lixingwang) [inline go.mod]: Please cleanup go.mod

File: trigger/wsclient/metadata.go [hunk 8]
```diff
@@ -3,6 +3,10 @@ package wsclient
 // Settings for the websocket client trigger
 type Settings struct {
 	URL string `md:"url,required"`
+	AllowInsecure bool   `md:"allowInsecure"`
+	CaCert        string `md:"caCert"`
+	QueryParams map[string]interface{} `md:"queryParams"`
```
Comment (lixingwang) [inline trigger/wsclient/metadata.go]: Change to map[string]string for query and header

File: trigger/wsclient/trigger.go [hunk 9]
```diff
@@ -38,32 +46,111 @@ type Trigger struct {
...
+	headerObject, err := coerce.ToObject(t.config.Settings["headers"])
+	if err != nil {
+		// queryparam is array as per FE
+		headerArray, err := coerce.ToArray(t.config.Settings["headers"])
+		if err != nil {
+			return err
+		}
+		for _, val := range headerArray {
+			qparamMap := val.(map[string]interface{})
+			header[qparamMap["parameterName"].(string)] = []string{qparamMap["value"].(string)}
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: I don't think we need this logic anymore base on our discussion, Let's cleanup

File: trigger/wsclient/trigger.go [hunk 10]
```diff
@@ -38,32 +46,111 @@ type Trigger struct {
...
+	} else {
+		dialer = *websocket.DefaultDialer
+	}
+
+	t.logger.Infof("dialing websocket endpoint[%s]...", urlstring)
+	conn, res, err := dialer.Dial(urlstring, header)
+	if err != nil {
+		if res != nil {
+			defer res.Body.Close()
+			body, err := ioutil.ReadAll(res.Body)
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: err handle

File: trigger/wsserver/server.go [hunk 11]
```diff
@@ -13,8 +14,13 @@ import (
 	"strings"
 	"sync"
 	"time"
+
+	"github.com/project-flogo/core/data/coerce"
+	"github.com/project-flogo/core/support/log"
 )
 
+var logger = log.ChildLogger(log.RootLogger(), "WSTrigger-Server")
```
Comment (lixingwang) [inline trigger/wsserver/server.go]: Please get logger from conext

File: trigger/wsserver/server.go [hunk 12]
```diff
@@ -72,12 +78,32 @@ func (s *Server) Start() error {
 
 	if s.enableTLS {
 		//TLS is enabled, load server certificate & key files
-		cer, err := tls.LoadX509KeyPair(s.serverCert, s.serverKey)
-		if err != nil {
-			fmt.Printf("Error while loading certificates - %v", err)
-			return err
+		logger.Info("Reading certificates")
```
Comment (lixingwang) [inline trigger/wsserver/server.go]: all logs we should get from context

**Review by lixingwang — CHANGES_REQUESTED** [review]

File: activity/ws/activity.go (hunk 3 above)
Comment (mbhagwat-tibco) [reply activity/ws/activity.go]: done.

File: activity/ws/activity.go (hunk 4 above)
Comment (mbhagwat-tibco) [reply activity/ws/activity.go]: done

File: activity/ws/activity.go (hunk 5 above)
Comment (mbhagwat-tibco) [reply activity/ws/activity.go]: done

File: activity/ws/metadata.go (hunk 6 above)
Comment (mbhagwat-tibco) [reply activity/ws/metadata.go]: done

File: trigger/wsclient/metadata.go (hunk 8 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/metadata.go]: done

File: trigger/wsclient/trigger.go (hunk 9 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: done

File: trigger/wsclient/trigger.go (hunk 10 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: done

File: trigger/wsserver/descriptor.json (hunk 1 above)
Comment (mbhagwat-tibco) [reply trigger/wsserver/descriptor.json]: done.

File: trigger/wsserver/server.go (hunk 12 above)
Comment (mbhagwat-tibco) [reply trigger/wsserver/server.go]: done

File: trigger/wsserver/server.go (hunk 11 above)
Comment (mbhagwat-tibco) [reply trigger/wsserver/server.go]: done

File: go.mod (hunk 7 above)
Comment (mbhagwat-tibco) [reply go.mod]: done

File: activity/ws/activity.go [hunk 13]
```diff
@@ -48,25 +54,214 @@ func (a *Activity) Metadata() *activity.Metadata {
...
+			if allowInsecure {
+				tlsconfig.InsecureSkipVerify = true
+			} else {
+				var cacertObj map[string]interface{}
+				if a.settings.CaCert != "" {
+					err = json.Unmarshal([]byte(a.settings.CaCert), &cacertObj)
+					if err != nil { //file path configured
+						certPool, err := getCerts(a.settings.CaCert)
+						if err != nil {
+							fmt.Printf("Error while loading client trust store - %v", err)
```
Comment (lixingwang) [inline activity/ws/activity.go]: Please use logger

File: activity/ws/activity.go [hunk 14]
```diff
@@ -48,25 +54,214 @@ func (a *Activity) Metadata() *activity.Metadata {
...
+			dialer = *websocket.DefaultDialer
+		}
+		ctx.Logger().Info("Creating new connection")
+		conn, res, err := dialer.Dial(builtURL, h)
+		if err != nil {
+			if res != nil {
+				defer res.Body.Close()
+				body, err1 := ioutil.ReadAll(res.Body)
+				if err1 != nil{
+					ctx.Logger().Infof("res code is %v error while reading response payload is %s ", res.StatusCode,  err1)
```
Comment (lixingwang) [inline activity/ws/activity.go]: Here should be error logs right? ctx.Logger().Errorf

File: trigger/wsclient/trigger.go [hunk 15]
```diff
@@ -42,28 +52,94 @@ func (*Factory) New(config *trigger.Config) (trigger.Trigger, error) {
...
+			tlsconfig.InsecureSkipVerify = true
+		} else {
+			caCertValue := t.settings.CaCert
+			var cacertObj map[string]interface{}
+			if caCertValue != "" {
+				err := json.Unmarshal([]byte(caCertValue), &cacertObj)
+				if err != nil { // filepath configured
+					certPool, err := getCerts(caCertValue)
+					if err != nil {
+						t.logger.Infof("Error while loading client trust store - %v", err)
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: error log here

File: trigger/wsclient/trigger.go [hunk 16]
```diff
@@ -42,28 +52,94 @@ func (*Factory) New(config *trigger.Config) (trigger.Trigger, error) {
...
+	t.logger.Debugf("dialing websocket endpoint[%s] with headers[%s]...", urlstring, header)
+	conn, res, err := dialer.Dial(urlstring, header)
 	if err != nil {
-		return fmt.Errorf("error while connecting to websocket endpoint[%s] - %s", url, err)
+		if res != nil {
+			defer res.Body.Close()
+			body, err1 := ioutil.ReadAll(res.Body)
+			if err1 != nil{
+				ctx.Logger().Infof("res code is %v error while reading response payload is %s ", res.StatusCode,  err1)
+			}
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go:127]: error log

File: trigger/wsclient/trigger.go [hunk 17]
```diff
@@ -42,28 +52,94 @@ func (*Factory) New(config *trigger.Config) (trigger.Trigger, error) {
...
+		}
+		return fmt.Errorf("error while connecting to websocket endpoint[%s] - %s", urlstring, err)
 	}
+
 	t.wsconn = conn
 	go func() {
+		defer func(){
+			err := conn.WriteMessage(websocket.CloseMessage, []byte("Sending close message while getting out of reading connection loop"))
+			if err != nil{
+				t.logger.Debugf("Received err [%s] while writing close message", err)
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: Maybe the error level is properly here.

File: trigger/wsclient/trigger.go [hunk 18]
```diff
@@ -91,6 +167,95 @@ func (t *Trigger) Start() error {
 
 // Stop stops the trigger
 func (t *Trigger) Stop() error {
+	err := t.wsconn.WriteMessage(websocket.CloseMessage, []byte("Closing connection while stopping trigger"))
+	if err != nil{
+		t.logger.Infof("Error received: [%s] while sending close message when Stopping Trigger", err)
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: error

**Review by lixingwang — APPROVED** [review]
Looks good to me after fixing the last small comments

File: activity/ws/activity.go (hunk 13 above)
Comment (milanbhagwat) [reply activity/ws/activity.go]: done

File: trigger/wsclient/trigger.go (hunk 18 above)
Comment (milanbhagwat) [reply trigger/wsclient/trigger.go]: converted to warn level as its not critical to be logged as error. done.

File: trigger/wsclient/trigger.go (hunk 17 above)
Comment (milanbhagwat) [reply trigger/wsclient/trigger.go]: converted to warn level as its not critical to be logged as error. done.

File: trigger/wsclient/trigger.go (hunk 16 above)
Comment (milanbhagwat) [reply trigger/wsclient/trigger.go:127]: done.

File: trigger/wsclient/trigger.go (hunk 15 above)
Comment (milanbhagwat) [reply trigger/wsclient/trigger.go]: done.

File: activity/ws/activity.go (hunk 14 above)
Comment (milanbhagwat) [reply activity/ws/activity.go]: done

**Review by vijaynalawade — APPROVED** [review]

File: trigger/wsserver/descriptor.json (hunk 2 above)
Comment (milanbhagwat) [reply trigger/wsserver/descriptor.json]: done on commit dated 1 March

File: trigger/wsserver/descriptor.json (hunk 1 above)
Comment (milanbhagwat) [reply trigger/wsserver/descriptor.json]: done on commit dated 24 march

**Pull request merged**